libtsuba/src/RDGManifest.cpp

```diff
@@ -0,0 +1,320 @@
+  tsuba::FileView fv;
+
+  if (auto res = fv.Bind(uri.string(), true); !res) {
+    return res.error();
+    auto latest_meta_uri = FindLatestManifestFile(uri);
+    if (!latest_meta_uri) {
+      return res.error();
+    }
+    if (auto res_latest = fv.Bind(latest_meta_uri.value().string(), true);
+        !res_latest) {
+      return res.error();
+    }
+  }
+  tsuba::RDGManifest meta(uri.DirName());
+  auto meta_res = katana::JsonParse<tsuba::RDGManifest>(fv, &meta);
+
+  if (!meta_res) {
+    return meta_res.error().WithContext("cannot parse {}", uri.string());
+  }
+
+  auto manifest_name = uri.BaseName();
+  auto view_name = ParseViewNameFromName(manifest_name);
+  auto view_args = ParseViewArgsFromName(manifest_name);
+
+  if (view_name) {
+    meta.set_viewtype(view_name.value());
+  }
+
+  if (view_args) {
+    meta.set_viewargs(view_args.value());
+  } else {
+    meta.set_viewargs(std::vector<std::string>());
+  }
+  return meta;
+}
+
+Result<RDGManifest>
+RDGManifest::Make(
+    const katana::Uri& uri, const std::string& view_type, uint64_t version) {
+  return MakeFromStorage(FileName(uri, view_type, version));
+}
+
+Result<RDGManifest>
+RDGManifest::Make(RDGHandle handle) {
+  return handle.impl_->rdg_meta();
+}
+
+Result<RDGManifest>
+RDGManifest::Make(const katana::Uri& uri) {
+  auto rdg_manifest = MakeFromStorage(uri);
+  return rdg_manifest;
+}
+
+std::string
+RDGManifest::PartitionFileName(
+    const std::string& view_type, uint32_t node_id, uint64_t version) {
+  KATANA_LOG_ASSERT(!view_type.empty());
+  return fmt::format(
+      "part_vers{:07d}_{}_node{:07d}", version, view_type, node_id);
+}
+
+katana::Uri
+RDGManifest::PartitionFileName(
+    const katana::Uri& uri, uint32_t node_id, uint64_t version) {
+  return uri.Join(
+      PartitionFileName(tsuba::gDefaultRDGViewType, node_id, version));
+}
+
+katana::Uri
+RDGManifest::PartitionFileName(
+    const std::string& view_type, const katana::Uri& uri, uint32_t node_id,
+    uint64_t version) {
+  KATANA_LOG_DEBUG_ASSERT(!IsMetaUri(uri));
+  return uri.Join(PartitionFileName(view_type, node_id, version));
+}
+
+katana::Uri
+RDGManifest::PartitionFileName(uint32_t host_id) const {
+  return RDGManifest::PartitionFileName(
+      view_specifier(), dir_, host_id, version());
+}
+
+std::string
+RDGManifest::ToJsonString() const {
+  // POSIX specifies that text files end in a newline
+  std::string s = json(*this).dump() + '\n';
+  return s;
+}
+
+// e.g., rdg_dir == s3://witchel-tests-east2/fault/simple/
+katana::Uri
+RDGManifest::FileName(
+    const katana::Uri& uri, const std::string& view_name, uint64_t version) {
+  KATANA_LOG_DEBUG_ASSERT(uri.empty() || !IsMetaUri(uri));
+  KATANA_LOG_ASSERT(!view_name.empty());
+  return uri.Join(
+      fmt::format("katana_vers{:07d}_{}.manifest", version, view_name));
+}
+
+// if it doesn't name a meta file, assume it's meant to be a managed URI
+bool
+RDGManifest::IsMetaUri(const katana::Uri& uri) {
```

Comment:
There are a couple of references to `Meta` that are now confusing. Should probably be `Manifest`